libs/accounts/rate-limit/src/lib/rate-limit.ts

```diff
@@ -0,0 +1,158 @@
+    const now = Date.now();
+
+    for (const rule of rules) {
+      // Sanitize key. Space is a decent replacement character, because it's an invalid character for
+      // ip addressees, emails, and user ids.
+      const blockedValue = opts[rule.blockingOn];
+
+      // Make sure the correct values were passed in as opts.
+      // If an action requires an email, ip, and/or uid, it must provided!
+      if (!blockedValue) {
+        throw new MissingOption(action, rule.blockingOn);
+      }
+
+      // Check to see if there are any blocks that currently exist in Redis.
+      // Create a new block and add set of openBlocks.
+      const attemptsKey = getKey('attempts', rule, blockedValue);
+      const blockKey = getKey('block', rule, blockedValue);
+
+      let block = JSON.parse((await this.redis.get(blockKey)) || 'null');
+      let attempts = 0;
+
+      if (block === null) {
+        attempts = await this.redis.incr(attemptsKey);
+        if (attempts === 1) {
+          await this.redis.expire(attemptsKey, rule.windowDurationInSeconds);
+        }
+
+        // If we've exceeded the max number of attempts, then create a block
+        if (attempts > rule.maxAttempts) {
+          block = {
+            action: rule.action,
+            startTime: now,
+            duration: rule.blockDurationInSeconds,
+            blockingOn: rule.blockingOn,
+            blockedValue: blockedValue,
+            reason: 'too-many-attempts',
+            attempts: attempts,
+          } satisfies BlockRecord;
+
+          await this.redis.set(blockKey, JSON.stringify(block));
+
+          await Promise.all([
+            this.redis.expire(blockKey, rule.blockDurationInSeconds),
+            this.redis.expire(attemptsKey, 0),
+          ]);
+        }
+      }
+
+      if (block !== null) {
+        openBlocks.push({
+          startTime: block.startTime,
+          duration: block.duration,
+          attempt: block.attempt,
+          retryAfter: calculateRetryAfter(now, block),
+          reason: 'too-many-attempts' as BlockReason,
+          action: rule.action,
+          blockingOn: rule.blockingOn,
+        });
+      }
+    }
+
+    // If blocks weren't found. Pick the block with the longest retry after.
+    if (openBlocks.length > 0) {
+      let block = openBlocks[0];
+      openBlocks.forEach((x) => {
+        if (x.retryAfter > block.retryAfter) {
+          block = x;
+        }
+      });
+
+      // TODO: Record metrics on this. We'd also like to improve these metrics and be able to
+      //  answer the following questions:
+      //    - What action was blocked, for what user, and on what condition (ie IP, uid, or email)
+      //    - How long the user will be blocked for
+      //    - Open question, do we also want to no blocks with shorter bans? Or just the block with largest
+      //      ban (ie the biggest retryAfter value).
+
+      return {
+        ...block,
+        openBlocks: openBlocks,
```

Comment:
What's the reason for returning all the open blocks?

Reply:
It's currently used in tests as a sanity check, which admittedly isn't a great reason to expose it. If we forgo these sanity checks, which is probably safe to do now, I don't think we actually need to expose this.

Reply:
I've made it so this is not loner exposed.